### ISSUE TITLE
difference between torch.bmm and a batch of torch.mm become larger when matrix dimension become smaller

### ISSUE THREAD

**Issue by budui** [pytorch/pytorch#47154]
## 🐛 Bug

difference between `torch.bmm` and a batch of `torch.mm` become larger when matrix dimension become smaller

## To Reproduce

```python
import torch


def diff(x1, x2):
    ans1 = x1 @ x2
    ans2 = []
    for i in range(x1.size(0)):
        ans2.append(x1[i] @ x2[i])
    ans2 = torch.stack(ans2)

    print('diff_mean =', (ans2 - ans1).abs().mean())
    print('diff_max =', (ans2 - ans1).abs().max())
    print()


if __name__ == '__main__':
    torch.manual_seed(0)
    x1 = torch.randn(32, 512, 64).cuda() * 1e4
    x2 = torch.randn(32, 64, 128).cuda() * 1e4

    diff(x1[:, :1, :], x2)
    diff(x1[:, :128 * 1, :], x2)
    diff(x1[:, :128 * 2, :], x2)
    diff(x1[:, :128 * 3, :], x2)
    diff(x1[:, :128 * 4, :], x2)
    diff(x1, x2)

```

## Expected behavior

difference  will be 0 at all times, or difference will be larger when matrix get larger.

## Actual behavior

```text
diff_mean = tensor(50.1438, device='cuda:0')
diff_max = tensor(512., device='cuda:0')

diff_mean = tensor(89.5844, device='cuda:0')
diff_max = tensor(1280., device='cuda:0')

diff_mean = tensor(89.4881, device='cuda:0')
diff_max = tensor(1792., device='cuda:0')

diff_mean = tensor(0., device='cuda:0')
diff_max = tensor(0., device='cuda:0')

diff_mean = tensor(0., device='cuda:0')
diff_max = tensor(0., device='cuda:0')

diff_mean = tensor(0., device='cuda:0')
diff_max = tensor(0., device='cuda:0')
```

## Environment

```text
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.6 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 430.34
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.18.5
[pip] pytorch-ignite==0.4.1
[pip] torch==1.7.0
[pip] torchaudio==0.7.0a0+ac17b64
[pip] torchvision==0.8.1
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] cudatoolkit               10.1.243             h6bb024c_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] ignite                    0.4.1                    py37_0    pytorch
[conda] mkl                       2020.1                      217    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-include               2020.1                      217    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.3.0            py37he904b0f_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.1.0            py37h23d657b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.1.1            py37h0573a6f_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.18.5           py37ha1c710e_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy-base                1.18.5           py37hde5b4d6_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] pytorch                   1.7.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchaudio                0.7.0                      py37    pytorch
[conda] torchvision               0.8.1                py37_cu101    pytorch
```

**Comment by ngimel** [comment]
You are comparing very large numbers, on the order of 10^8. The differences you are observing stem from floating point accuracy. cublas can use different kernels that sum the numbers in the different orders, there's no guarantee that results you are getting will exactly match single matrix multiplication or that the error will grow with the matrix size.
